posts/Jan/2022-04-29-mapping-boolean-types-in-hibernate-orm-6.adoc

```diff
@@ -0,0 +1,149 @@
+
+NOTE: in impending ORM 6.1 release, a new `@ConverterRegistration` was added to make this much easier. It allows modifying certain aspects of already defined `AttributeConverters`, such as the application domain (`autoApply`) or change the data type the conversion should be applied on to another compatible data type (e.g. a subclass of the original data type).
+
+.Customizing built-in AttributeConverters
+====
+[source, JAVA, indent=0]
+----
+// this will apply the converter to all Boolean attributes
+@Basic
+@ConverterRegistration( converter = org.hibernate.type.YesNoConverter.class, autoApply = true )
+boolean convertedYesNo;
+
+----
+====
+
+If a more unusual conversion is needed to map this kind of attribute to the database, one can also implement a custom `AttributeConverter` - check link:{user-guide-url}#basic-jpa-convert[AttributeConverters] for more info.
+
+Such a custom `AttributeConverter` can also be applied to all the Boolean attributes, by setting its `autoApply` attribute to true.
+
+.Defining an AttributeConverter as global
+====
+[source, JAVA, indent=0]
+----
+
+    @Converter(autoApply = true)
+    public class CustomBooleanConverter implements AttributeConverter<Boolean, Character> {
+        ...
+    }
+
+----
+====
+
+NOTE: for this to work, the converter needs to be registered in the `persistence.xml`, since Hibernate doesn't auto-discover them (yet)
+====
+[source, XML, indent=0]
+----
+    <persistence-unit ...>
+        ...
+        <class>CustomBooleanConverter</class>
+        ...
+    </persistence-unit>
+----
+====
+
+
+== Hibernate extensions
+
+=== `@JavaType`
+
+In general the standard `BooleanJavaType` will cover most cases of possible mappings for a boolean attribute. If not, one can always implement a custom extension of the new `JavaType` interface that was introduced in ORM 6, to cover any special needs. This custom `JavaType` can then be applied that either locally to a specific attribute with the new `@JavaType` annotation, or globally, by annotating the custom `JavaType` with the new `@JavaTypeRegistry` annotation.
+See the user guide section on link:{user-guide-url}#basic-mapping-composition[compositional basic mapping] for more info on how to do this.
+
+=== `@JdbcTypeCode / @JdbcType`
+
+If one wants to map a specific boolean attribute to something different from the default for the current database, then the easiest way to achieve this is to annotate that attribute with the desired `@JdbcTypeCode`, or with `@JdbcType` and the desired `JdbcType` implementor. In the latter case one can either use an implementation provided by Hibernate that accounts for the desired transformation, or provide a custom implementation.
+
+.Apply a specific Jdbc Type locally to a Boolean
+====
+[source, JAVA, indent=0]
+----
+    // this Boolean attribute will be converted to a CHAR in the database
+    @JdbcTypeCode(SqlTypes.CHAR)
```

Comment:
Have you verified that this actually works?

Reply:
This tells Hibernate the type of JDBC type to use, not how to perform a conversion from boolean -> char.

What gets stored when you do this?

Reply:
in H2: CHAR(1) CHECK ("B" IN('N', 'Y'))
So, Hibernate does magic even beyond it's makers' knowledge ... :D

Reply:
Might be worth mentioning in the blog post, that `Y`/`N` is the "default" representation as per `BooleanJavaType`

Reply:
> YesNoConverter:: encodes boolean values as 'Y' or 'N'